cmd/nodecmd/upgrade.go

```diff
@@ -0,0 +1,224 @@
+// getNodesToBeUpgraded gets the node versions of all nodes in cluster clusterName and checks which
+// nodes needs to have Avalanche Go & SubnetEVM upgraded. It first checks the subnet EVM version -
+// it will install the newest subnet EVM version and install the latest avalanche Go that is still compatible with the Subnet EVM version
+// if the node is not tracking any subnet, it will just install latestAvagoVersion
+// sample return result map[aws_node_i-006e0fbf1df7e2198:avalanchego=v1.10.13,subnet-evm=n8Anw9kErmgk7KHuuxRag42nR4xhhBeePH2MkwSxfGNPiaCJs:v0.5.7
+// aws_node_i-06af8a2ebbeb1bf01:avalanchego=v1.10.13,subnet-evm=n8Anw9kErmgk7KHuuxRag42nR4xhhBeePH2MkwSxfGNPiaCJs:v0.5.7]
+func getNodesToBeUpgraded(clusterName string) (map[string]string, error) {
+	latestAvagoVersion, err := app.Downloader.GetLatestReleaseVersion(binutils.GetGithubLatestReleaseURL(
+		constants.AvaLabsOrg,
+		constants.AvalancheGoRepoName,
+	))
+	if err != nil {
+		return nil, err
+	}
+	latestSubnetEVMVersion, err := app.Downloader.GetLatestReleaseVersion(binutils.GetGithubLatestReleaseURL(
+		constants.AvaLabsOrg,
+		constants.SubnetEVMRepoName,
+	))
+	if err != nil {
+		return nil, err
+	}
+	rpcVersion, err := vm.GetRPCProtocolVersion(app, models.SubnetEvm, latestSubnetEVMVersion)
+	if err != nil {
+		return nil, err
+	}
+	ansibleNodeIDs, err := ansible.GetAnsibleHostsFromInventory(app.GetAnsibleInventoryDirPath(clusterName))
+	if err != nil {
+		return nil, err
+	}
+	failedNodes := []string{}
+	nodeErrors := []error{}
+	nodesToUpgrade := make(map[string]string)
+	for _, host := range ansibleNodeIDs {
+		if err := app.CreateAnsibleStatusFile(app.GetAvalancheGoJSONFile()); err != nil {
+			failedNodes = append(failedNodes, host)
+			nodeErrors = append(nodeErrors, err)
+			continue
+		}
+		if err := ansible.RunAnsiblePlaybookCheckAvalancheGoVersion(app.GetAnsibleDir(), app.GetAvalancheGoJSONFile(), app.GetAnsibleInventoryDirPath(clusterName), host); err != nil {
+			failedNodes = append(failedNodes, host)
+			nodeErrors = append(nodeErrors, err)
+			continue
+		}
+		vmVersions, err := parseNodeVersionOutput(app.GetAvalancheGoJSONFile())
+		if err != nil {
+			failedNodes = append(failedNodes, host)
+			nodeErrors = append(nodeErrors, err)
+			continue
+		}
+		currentAvalancheGoVersion := vmVersions[constants.PlatformKeyName]
+		var avalancheGoVersionToUpdateTo string
+		var subnetEVMVersionToUpdateTo string
+		for vmName, vmVersion := range vmVersions {
+			// when calling info.getNodeVersion, this is what we get
+			// "vmVersions":{"avm":"v1.10.12","evm":"v0.12.5","n8Anw9kErmgk7KHviddYtecCmziLZTphDwfL1V2DfnFjWZXbE":"v0.5.6","platform":"v1.10.12"}},
+			// we need to get the VM ID of the subnets that the node is currently validating, in the example above it is n8Anw9kErmgk7KHviddYtecCmziLZTphDwfL1V2DfnFjWZXbE
+			if vmName != constants.PlatformKeyName && vmName != constants.EVMKeyName && vmName != constants.AVMKeyName {
+				if vmVersion != latestSubnetEVMVersion {
```

Comment:
you need to check if the vm is custom vm or subnet-evm

Reply:
you can either search the remote sidecar or the local sidecar (probably the remote sidecar?) but a check is needed.
for custom vms, another thing shoud be made in this same command. (another PR)
what do you think is best? remote sidecar or local sidecar? probably remote. can we have a new CLI command that
provides all the needed info for all the subnets so as to search for the vm? and call that command. maybe we can use one of the current subnet cli commands like list.

Reply:
I believe it is better to call a remote CLI command and parse the output locally than to get the remote sidecars